### PR TITLE
add constructor options for timeout, stop using transport.DialTimeout

### PR DESCRIPTION
**Pull request by marten-seemann**

### CODE REVIEW

File: dial_test.go [hunk 1]
```diff
@@ -161,8 +156,9 @@ func newSilentPeer(t *testing.T) (peer.ID, ma.Multiaddr, net.Listener) {
 func TestDialWait(t *testing.T) {
 	t.Parallel()
 
-	ctx := context.Background()
```
Comment (vyzo) [inline dial_test.go:164]: again here, why are we deopping this context?

Comment (marten-seemann) [reply dial_test.go:164]: Because it didn't do anything, it's a `context.Background()`.

Comment (vyzo) [reply dial_test.go:164]: ok.

**Review by vyzo — COMMENTED** [review]
how will this option get propagated?

**Comment by marten-seemann** [comment]
Like this: https://github.com/libp2p/go-libp2p/pull/1271

**Review by vyzo — APPROVED** [review]
ok, looks good.

**Pull request merged**